added snap schedule form

ivoalmeida opened the pull request:
Fixes: https://tracker.ceph.com/issues/63827

Added create snapshot schedule endpoint and modal form to create snapshot schedules and retention policies at the same time.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

[Screencast from 2024-01-23 13-40-43.webm](https://github.com/ceph/ceph/assets/1071297/3740a387-d4e7-422f-91eb-6796c2ac7b24)

ivoalmeida commented:
I will create a separate PR for sub-volumes

cloudbehl commented:
Thanks for PR. Few general comments.

1. Maybe calling `Directory` ->  `Directory Path`. Can we show the loading icon while the paths are loading?
2. I think we can have multiple schedule per path as well. So the validation needs to be if the path and the schedule is same then user can't create it. 
3. The time and date we choose is UTC. We need to let user know about it and also set the default value to UTC date and time. Long term maybe we can allow them to choose the timezone as well.
4. Can we pre select today's UTC date as well?
5. Can we just deal in Hours and minutes in time and not care about seconds.
6. Lets call `hour -> Hourly` ,`day -> Daily`, `week -> Weekly`.  Also I think the default value can be set to Daily
7. Something I'm finding confusing, in snapshot we say 1 hour, 2 hour or 4 hour. As a user it confuses me that if its 2 snapshot per hour or 4 snapshot per hour but actually it is 1 snapshot every 2 hours or 4 hours. 
8. Retention is not mandatory, so maybe we can set one value as forever and disable the text box. If other value is selected we can ask to add the value. Also we need to disable the `+` button based on the values. 
9. With `latest snapshot`, the list page show value 1N. As a user its not understandable so needs a fix in list page.

Pegonzal reviewed (commented):
Looks really good @ivoalmeida! 

I'll leave some feedback and concerns I have:

1. My biggest concern is the `directory` field. As @cloudbehl already suggested a spinner to indicate is loading would be a nice addition, but I think it has to go a little further than that. Two main issues I find:
     -  Long loading times. On clusters with many (thousands) of subvolumes it could take some significant time to load them all

     - Usability. It can be hard to find the item you want in a dropdown of thousands (that can even have similar names), example with just a few hundred: 
[directory_dropdown.webm](https://github.com/ceph/ceph/assets/106810838/decf68c3-586c-4f31-96e5-af3ead5b944e)

        _Proposal_: Have the user first select the subvolume group and then load and list the volumes in that selected subvolume group if needed. I don't think this will solve the problem completly, but it would help. Some kind of pagination and filtering might be a good addition in the future.

2. Mark each field as required if applicable.

3. When selecting the `Schedule` and `Retention policy` values, a green check comes up, if it is not of much effort we could get rid of it as it doesn't add anything other than a wierd spacing in the form.

4. On the UTC time and also answering to @cloudbehl 's comment 
     > The time and date we choose is UTC.
    
    Even though the snapshot scheduler is working with UTC times, could we translate them in the Dashboard? That way the user could work with it's local time.
    This might be too troublesome to be converting back and forth UTC times to local, but I think it's something to consider
4. Don't forget to prepend the commit and PR title with `mgr/dashboard: `

ivoalmeida commented:
> * Maybe calling `Directory` ->  `Directory Path`. Can we show the loading icon while the paths are loading?
> * I think we can have multiple schedule per path as well. So the validation needs to be if the path and the schedule is same then user can't create it.
> * The time and date we choose is UTC. We need to let user know about it and also set the default value to UTC date and time. Long term maybe we can allow them to choose the timezone as well.
> * Can we pre select today's UTC date as well?
> * Can we just deal in Hours and minutes in time and not care about seconds.
> * Lets call `hour -> Hourly` ,`day -> Daily`, `week -> Weekly`.  Also I think the default value can be set to Daily
> * Something I'm finding confusing, in snapshot we say 1 hour, 2 hour or 4 hour. As a user it confuses me that if its 2 snapshot per hour or 4 snapshot per hour but actually it is 1 snapshot every 2 hours or 4 hours.
> * Retention is not mandatory, so maybe we can set one value as forever and disable the text box. If other value is selected we can ask to add the value. Also we need to disable the `+` button based on the values.
> * With `latest snapshot`, the list page show value 1N. As a user its not understandable so needs a fix in list page.

Hi @cloudbehl 

I have implemented changes to address your comments.

1. I added the loading icon. I didn't change because I am not sure if this is an improvement. 
2. I removed the validation from the directory and added to the frequency and retention frequency units. According to my tests, you can have more than 1 frequency and retention policy per directory as long as the frequency unit is not the same.
3. I added a helper message letting the user know about the UTC. I am not sure about the copy though. I copied from the Ceph documentation. Regarding the time zone, I tried adding it in the CLI using one example from the documentation  and got an error.
4. Added the initial date and time in UTC.
5. I removed it.
6. Not sure, what I am supposed to do here. 
7. I made retention optional. I cannot set retention as 'forever'. I am not sure how long a snapshot is kept for if not retention policy is assigned to it. I couldn't find a clear answer from the documentation. I don't understand this> _If other value is selected we can ask to add the value. Also we need to disable the `+` button based on the values._ Can you please elaborate?
8. I am planning a PR just to add the copies. No copies for the values returned from the backend have been added yet other then the frequency units.

ivoalmeida commented:
> Looks really good @ivoalmeida!
> 
> I'll leave some feedback and concerns I have:
> 
> 1. My biggest concern is the `directory` field. As @cloudbehl already suggested a spinner to indicate is loading would be a nice addition, but I think it has to go a little further than that. Two main issues I find:
>    
>    * Long loading times. On clusters with many (thousands) of subvolumes it could take some significant time to load them all
>    * Usability. It can be hard to find the item you want in a dropdown of thousands (that can even have similar names), example with just a few hundred:
>      [directory_dropdown.webm](https://github.com/ceph/ceph/assets/106810838/decf68c3-586c-4f31-96e5-af3ead5b944e)
>      _Proposal_: Have the user first select the subvolume group and then load and list the volumes in that selected subvolume group if needed. I don't think this will solve the problem completly, but it would help. Some kind of pagination and filtering might be a good addition in the future.
> 2. Mark each field as required if applicable.
> 3. When selecting the `Schedule` and `Retention policy` values, a green check comes up, if it is not of much effort we could get rid of it as it doesn't add anything other than a wierd spacing in the form.
> 4. On the UTC time and also answering to @cloudbehl 's comment
>    > The time and date we choose is UTC.
>    
>    
>    Even though the snapshot scheduler is working with UTC times, could we translate them in the Dashboard? That way the user could work with it's local time.
>    This might be too troublesome to be converting back and forth UTC times to local, but I think it's something to consider
> 5. Don't forget to prepend the commit and PR title with `mgr/dashboard: `

Hi @Pegonzal 

1. I added the loading icon there. Aside from that I am planning to make optimisations to how the directories are retrieved but that will be handled in a separate PR. I have talked about it with Nizam and Ankush already while you were on holidays last year. I'll take your suggestions into account when I am implementing this. Thank you.
2. Done
3. They are global styles present in _forms.scss. 
4. I'll let Ankush answer you
5. Done

cloudbehl commented:
Few more things I observed, while testing
1. The toast notification says `Unknown task executed`.
2. If the schedule is not created then also the validation is coming.
3. Different schedule can be added for 1h, 2h, 3h, but the validation stops as it only allows 1 hourly schedule, 1 daily and 1 weekly.
4. Once we click on + button of adding retention I can't go back and delete it if I don't want to add it.

ivoalmeida commented:
> Few more things I observed, while testing
> 
> 1. The toast notification says `Unknown task executed`.
> 2. If the schedule is not created then also the validation is coming.
> 3. Different schedule can be added for 1h, 2h, 3h, but the validation stops as it only allows 1 hourly schedule, 1 daily and 1 weekly.
> 4. Once we click on + button of adding retention I can't go back and delete it if I don't want to add it.

1. Fixed. This got messed up after my latest ```main rebase``` and I missed this.
2. I think you are referring to the validation kicking in before you make any changes to the schedule, right? If so, I removed that. I wasn't sure about it myself.
3. Fixed. That was a wrong assumption because with the retention you can only ever have one frequency unit but for schedule you can have same frequency unit with different interval number.
4. I added this functionality.

cloudbehl reviewed (approved):
LGTM, the functionality is working all good now.

@ivoalmeida can you make sure the tests are passing.

nizamial09 commented:
@ivoalmeida you will need to rebase this PR against the main to fix the doc issues

nizamial09 commented on src/pybind/mgr/dashboard/controllers/cephfs.py:
```diff
@@ -922,8 +922,8 @@ def list(self, fs: str, path: str = '/', recursive: bool = True):
-                                                     current_path, fs, 'plain')
-            output.append(json.loads(status_out))
+                                                     current_path, fs, None, None, 'json')
+            output = output + json.loads(status_out)
```
it could be because you intialized output as a list and then here you are appending the json string to it so essentially the type of the variable is being changed here.

ivoalmeida replied:
I had to make this change in order for it to work with snap schedule of the same path. The output from the module is very strange and it doesn't match  the output of the CLI command. Loading the data from the module as 'plan' caused me loads of problems.

ivoalmeida commented:
> @ivoalmeida I see this error in mypy
> 
> ```
> dashboard/__init__.py:60: note: ... from here:
> dashboard/controllers/cephfs.py: note: In member "list" of class "CephFSSnapshotSchedule":
> dashboard/controllers/cephfs.py:920: error: Need type annotation for "output" (hint: "output: List[<type>] = ...")
> Found 1 error in 1 file (checked 32 source files)
> mypy: exit 1 (64.27 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr> mypy --config-file=../../mypy.ini -m alerts -m balancer -m cephadm -m crash -m dashboard -m devicehealth -m diskprediction_local -m hello -m influx -m iostat -m localpool -m mds_autoscaler -m mgr_module -m mgr_util -m mirroring -m nfs -m orchestrator -m pg_autoscaler -m progress -m prometheus -m rbd_support -m rgw -m rook -m snap_schedule -m selftest -m stats -m status -m telegraf -m telemetry -m test_orchestrator -m volumes -m zabbix pid=3065128
> mypy: FAIL ✖ in 1 minute 45.75 seconds
> ```
> 
> while you are fixing that, can you also prefix the commit message with `mgr/dashboard: `. thanks.

which command did you run to get this message? I ran all linting and tests and I got no warning like this.

![image](https://github.com/ceph/ceph/assets/1071297/ac0224fd-495a-47d0-88e0-8afb6c3759c8)

nizamial09 commented:
> > @ivoalmeida I see this error in mypy
> > ```
> > dashboard/__init__.py:60: note: ... from here:
> > dashboard/controllers/cephfs.py: note: In member "list" of class "CephFSSnapshotSchedule":
> > dashboard/controllers/cephfs.py:920: error: Need type annotation for "output" (hint: "output: List[<type>] = ...")
> > Found 1 error in 1 file (checked 32 source files)
> > mypy: exit 1 (64.27 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr> mypy --config-file=../../mypy.ini -m alerts -m balancer -m cephadm -m crash -m dashboard -m devicehealth -m diskprediction_local -m hello -m influx -m iostat -m localpool -m mds_autoscaler -m mgr_module -m mgr_util -m mirroring -m nfs -m orchestrator -m pg_autoscaler -m progress -m prometheus -m rbd_support -m rgw -m rook -m snap_schedule -m selftest -m stats -m status -m telegraf -m telemetry -m test_orchestrator -m volumes -m zabbix pid=3065128
> > mypy: FAIL ✖ in 1 minute 45.75 seconds
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > while you are fixing that, can you also prefix the commit message with `mgr/dashboard: `. thanks.
> 
> which command did you run to get this message? I ran all linting and tests and I got no warning like this.
> 
> ![image](https://private-user-images.githubusercontent.com/1071297/300753790-ac0224fd-495a-47d0-88e0-8afb6c3759c8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2MDgwNjgsIm5iZiI6MTcwNjYwNzc2OCwicGF0aCI6Ii8xMDcxMjk3LzMwMDc1Mzc5MC1hYzAyMjRmZC00OTVhLTQ3ZDAtODhlMC04YWZiNmMzNzU5YzgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMzBUMDk0MjQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmQ2YTQ2YjcwNmEwYWM0ZmExYjU3NTAxODdkYTU2Mzk3NjE3M2NkMjE2ZDllMDE3ZTYzYTJlYjIxMGY5MTYzNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2GKTbttCo-i0VxnImS8cecpvnwsck7S1GbV9FKKxdGs)

its the mypy one. using ceph-dev
```
Check dashboard python code with mypy:
# Enable mypy check in .env file:
CHECK_MYPY=1

docker-compose run --rm ceph /docker/ci/sanity-checks.sh run_mypy

# Only 1 file:
docker-compose run --rm ceph /docker/ci/sanity-checks.sh run_mypy src/pybind/mgr/dashboard/controllers/rgw.py

```

ivoalmeida commented:
> docker-compose run --rm ceph /docker/ci/sanity-checks.sh run_mypy

I made the change but could not run the command. I get this error:
`Running mypy...
cp: cannot stat './rook-client-python/rook_client': No such file or directory
`

ivoalmeida commented:
jenkins test windows

phlogistonjohn commented:
Hi all,
A change from this PR seems to break running mypy under python 3.7.
```
py37-mypy: commands[0]> mypy --config-file=../../mypy.ini -m alerts -m balancer -m cephadm -m crash -m dashboard -m devicehealth -m diskprediction_local -m hello -m influx -m iostat -m localpool -m mds_autoscaler -m mgr_module -m mgr_util -m mirroring -m nfs -m orchestrator -m pg_autoscaler -m progress -m prometheus -m rbd_support -m rgw -m rook -m selftest -m smb -m snap_schedule -m stats -m status -m telegraf -m telemetry -m test_orchestrator -m volumes -m zabbix
dashboard/plugins/feature_toggles.py:9: note: In module imported here,
dashboard/module.py:55: note: ... from here,
dashboard/__init__.py:60: note: ... from here:
dashboard/controllers/cephfs.py: note: In member "list" of class "CephFSSnapshotSchedule":
dashboard/controllers/cephfs.py:955: error: "list" is not subscriptable, use "typing.List" instead  [misc]
Found 1 error in 1 file (checked 33 source files)
```

AFAIK ceph still is supporting python 3.6 (RHEL/Centos 8) so we should probably fix this since it's a very easy one.
